### PR TITLE
[Rtf] Refactored the encoding code.

### PR DESCRIPTION
**Pull request by luyi404**
**What this PR does**:

- For the changes in the dependent dubbogo/grpc-go (https://github.com/dubbogo/grpc-go/pull/5), the code of the Codec was refactored.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

**Comment by sheny1xuan** [comment]
Plz run `go fmt` for these code. @luyi404

**Comment by AlexStocks** [comment]
> Plz run `go fmt` for these code. @luyi404

Yeah.

**Comment by AlexStocks** [comment]
hi @luyi404 , if u do not wanna add comment for some funcs such as "Name()", please just add "// nolint" as comment.

**Comment by AlexStocks** [comment]
And pls check the ci failure after u summit every commit.

File: pkg/common/encoding/javaType/javaType.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
+package javaType
```
Comment (xavier-niu) [inline pkg/common/encoding/javaType/javaType.go]: Please rename `javaType` to `java_type` for both package and directory.

**Comment by justxuewei** [comment]
Why you remove all examples from the project?

**Comment by luyi404** [comment]
> Why you remove all examples from the project?

The example codes depend on outdated versions of both dubbo-go and triple. Besides, I need to remove the reliability of dubbo-go in triple to pass the CI.

File: pkg/common/encoding/javaType/javaType.go (hunk 1 above)
Comment (luyi404) [reply pkg/common/encoding/javaType/javaType.go]: done

**Comment by justxuewei** [comment]
> The example codes depend on outdated versions of both dubbo-go and triple. Besides, I need to remove the reliability of dubbo-go in triple to pass the CI.

OK, please ignoring that, the examples are duplicated with dubbo-go-samples. Please ensure that all samples in dubbo-go-samples works correctly.

File: pkg/common/encoding/java_type/java_type.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	"reflect"
+	"time"
+)
+
+import (
+	hessian "github.com/apache/dubbo-go-hessian2"
+)
+
+// GetArgType is copied from hessian, to get java class type of interface
+func GetArgType(v interface{}) string {
```
Comment (xavier-niu) [inline pkg/common/encoding/java_type/java_type.go:30]: I am thinking here should call hessian API instead of copying & pasting. You may submit a PR for hessian2 to wrap an exported method.

Comment (xavier-niu) [reply pkg/common/encoding/java_type/java_type.go:30]: If this function is updated by hessian2, it will cause some problems unexpectedly.

File: pkg/common/encoding/hessian/hessian.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/dubbogo/triple/pkg/common/encoding"
+	"github.com/dubbogo/triple/pkg/common/encoding/tools"
+)
+
+func init() {
+	encoding.RegisterCodec(pbwrapper.NewPBWrapperTwoWayCodec("hessian2", NewHessianCodec(), raw_proto.NewProtobufCodec()))
+}
+
+// HessianCodeC is the hessian impl of Codec interface
+type HessianCodeC struct{}
```
Comment (xavier-niu) [inline pkg/common/encoding/hessian/hessian.go:19]: Rename `HessianCodeC` to `HessianCodec`

File: pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+func (x *TripleRequestWrapper) GetArgs() [][]byte {
+	if x != nil {
+		return x.Args
+	}
+	return nil
+}
+
+func (x *TripleRequestWrapper) GetArgTypes() []string {
+	if x != nil {
```
Comment (xavier-niu) [inline pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go:97]: I think we don't need to test if `x` is nil or not, because if `x` is nil, calling a method will cause panic.

File: pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+	return mi.MessageOf(x)
+}
+
+// Deprecated: Use TripleResponseWrapper.ProtoReflect.Descriptor instead.
+func (*TripleResponseWrapper) Descriptor() ([]byte, []int) {
+	return file_triple_wrapper_proto_rawDescGZIP(), []int{1}
+}
+
+func (x *TripleResponseWrapper) GetSerializeType() string {
+	if x != nil {
```
Comment (xavier-niu) [inline pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go:146]: same above

File: pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go (hunk 4 above)
Comment (xavier-niu) [reply pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go:97]: Btw, is this file generated by some tools?

File: pkg/common/encoding/tools/tools.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	if reflect.TypeOf(out).Kind() != reflect.Ptr {
+		return perrors.Errorf("@out should be a pointer")
+	}
+
+	inValue := hessian.EnsurePackValue(in)
+	outValue := hessian.EnsurePackValue(out)
+
+	outType := outValue.Type().String()
+	if outType == "interface {}" || outType == "*interface {}" {
```
Comment (xavier-niu) [inline pkg/common/encoding/tools/tools.go:48]: Comparing the string of type is very tricky, I am thinking you should use both `Type()` and dereference method.

Comment (xavier-niu) [reply pkg/common/encoding/tools/tools.go:48]: You may refer to https://stackoverflow.com/questions/54012738/dereference-struct-pointer-and-access-fields-with-reflection.

File: pkg/common/encoding/java_type/java_type.go (hunk 2 above)
Comment (luyi404) [reply pkg/common/encoding/java_type/java_type.go:30]: This file comes from a previous modification to grpc-go that was tested in dubbo-go-example.

File: pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go (hunk 4 above)
Comment (luyi404) [reply pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go:97]: > Btw, is this file generated by some tools?

yes, its generated by `protoc` from file `triple_wrapper.proto`.


> I think we don't need to test if `x` is nil or not, because if `x` is nil, calling a method will cause panic.

I tested that calling a method from a nil pointer with an explicit context type will not cause a panic in golang, unless using its "member variables".

Comment (xavier-niu) [reply pkg/common/encoding/proto_wrapper_api/triple_wrapper.pb.go:97]: Ok, got it.

File: pkg/common/encoding/java_type/java_type.go (hunk 2 above)
Comment (xavier-niu) [reply pkg/common/encoding/java_type/java_type.go:30]: I am saying that this function may be updated by hessian2 in the future, you may refer to https://github.com/apache/dubbo-go-hessian2/blob/master/request.go, and I didn't see the difference here. You could submit a PR to rename the `getArgType` method to `GetArgType`. After the PR is merged, bump the version and call it directly.

File: pkg/triple/dubbo3_server.go [hunk 7]
```diff
@@ -154,43 +155,110 @@ func (h *GenericCodec) UnmarshalRequest(data []byte) ([]interface{}, error) {
...
+				for k, v := range outerAttachment {
+					if str, ok := v.(string); ok {
+						responseAttachment[k] = []string{str}
+					} else if strs, ok := v.([]string); ok {
+						responseAttachment[k] = strs
 					}
-				},
-			},
-		},
+					// todo deal with unsupported attachment
```
Comment (xavier-niu) [inline pkg/triple/dubbo3_server.go:240]: Add a warning log please.

**Comment by justxuewei** [comment]
lgtm, thanks for @luyi404.

**Review by AlexStocks — APPROVED** [review]

**Comment by AlexStocks** [comment]
@沈逸飞 继续后续工作
